### PR TITLE
Change default of deferServletRequestListenerDestroyOnError for Servlet-5.0

### LINKED ISSUE #12633: Review default of deferServletRequestListenerDestroyOnError for Servlet 5
See https://github.com/OpenLiberty/open-liberty/issues/9929#issuecomment-644736561

### PR DESCRIPTION
**Pull request by volosied**
fixes #12633

This is a draft and still needs further review.  

If we decide to proceed further, then we will need to schedule a POC to approve this change.

### CODE REVIEW

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer.servlet.3.1_fat

File: dev/com.ibm.ws.webcontainer/src/com/ibm/wsspi/webcontainer/WCCustomProperties.java [hunk 1]
```diff
@@ -756,7 +756,7 @@ public static void setCustomPropertyVariables() {
         if (SERVLET31_PRIVATE_BUFFERSIZE_FOR_LARGE_POST_DATA < 1) SERVLET31_PRIVATE_BUFFERSIZE_FOR_LARGE_POST_DATA = Integer.MAX_VALUE/16;
        
         //Start 8.5.5.5 CD
-        DEFER_SERVLET_REQUEST_LISTENER_DESTROY_ON_ERROR =  Boolean.valueOf(WebContainer.getWebContainerProperties().getProperty("com.ibm.ws.webcontainer.deferservletrequestlistenerdestroyonerror")).booleanValue(); //PI26908
+        DEFER_SERVLET_REQUEST_LISTENER_DESTROY_ON_ERROR =  Boolean.valueOf(WebContainer.getWebContainerProperties().getProperty("com.ibm.ws.webcontainer.deferservletrequestlistenerdestroyonerror", "true")).booleanValue(); //PI26908
```
Comment (volosied) [inline dev/com.ibm.ws.webcontainer/src/com/ibm/wsspi/webcontainer/WCCustomProperties.java]: Only used for FAT build. Will remove this line and un-commented the commented portions below if this change is approved for servlet 5.0.

**Review by pmd1nh — APPROVED** [review]

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer.servlet.3.1_fat

**Comment by pnicolucci** [comment]
@volosied what's the current status of this PR?

**Comment by volosied** [comment]
Still on hold because of of a failing CTS CDI test. I recreated the app, but I still need to dig further into this change.

I will follow on this one as well.

**Comment by volosied** [comment]
The failing test is https://github.com/eclipse-ee4j/cdi-tck/blob/0f9f0a484f10903e484b676e6bcea62ec97e91e3/web/src/main/java/org/jboss/cdi/tck/tests/context/conversation/determination/ConversationDeterminationTest.java

This exception occurs during the test. 

```
WELD-000315: Failed to acquire conversation lock in 1,000 ms for Conversation with id: foo

 Exception thrown by application class 'org.jboss.weld.contexts.AbstractConversationContext.initialize:239'
org.jboss.weld.contexts.BusyConversationException: WELD-000322: Conversation lock timed out: foo
at org.jboss.weld.contexts.AbstractConversationContext.initialize(AbstractConversationContext.java:239)
at org.jboss.weld.module.web.context.http.LazyHttpConversationContextImpl.initialize(LazyHttpConversationContextImpl.java:86)
at org.jboss.weld.module.web.context.http.LazyHttpConversationContextImpl.checkContextInitialized(LazyHttpConversationContextImpl.java:121)
at org.jboss.weld.contexts.AbstractConversationContext.getCurrentConversation(AbstractConversationContext.java:508)
at org.jboss.weld.bean.builtin.ConversationBean.newInstance(ConversationBean.java:31)
at org.jboss.weld.bean.builtin.ConversationBean.newInstance(ConversationBean.java:16)
at org.jboss.weld.bean.builtin.AbstractDecorableBuiltInBean.create(AbstractDecorableBuiltInBean.java:53)
at org.jboss.weld.contexts.AbstractContext.get(AbstractContext.java:96)
at org.jboss.weld.bean.ContextualInstanceStrategy$DefaultContextualInstanceStrategy.get(ContextualInstanceStrategy.java:100)
at org.jboss.weld.bean.ContextualInstance.get(ContextualInstance.java:50)
at org.jboss.weld.bean.proxy.ContextBeanInstance.getInstance(ContextBeanInstance.java:102)
at org.jboss.weld.bean.proxy.ProxyMethodHandler.invoke(ProxyMethodHandler.java:105)
at jakarta.enterprise.context.Conversation$Proxy$_$$_WeldClientProxy.getId(Unknown Source)
at org.jboss.cdi.tck.tests.context.conversation.determination.QuxAsyncListener.checkSameConversationActive(QuxAsyncListener.java:65)
at org.jboss.cdi.tck.tests.context.conversation.determination.QuxAsyncListener.onStartAsync(QuxAsyncListener.java:61)
at com.ibm.ws.webcontainer.async.AsyncListenerEntry.invokeOnStartAsync(AsyncListenerEntry.java:112)
at [internal classes]
at org.jboss.cdi.tck.tests.context.conversation.determination.AsyncFooServlet.service(AsyncFooServlet.java:60)
at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:587)
at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1258)
at [internal classes]
```

The test tests four endpoints. 
- foo-async?action=complete&cid=foo
- foo-async?action=timeout&cid=foo
- foo-async?action=error&cid=foo  
- foo-async?action=loop&cid=foo 

During `foo-async?action=error&cid=foo`, an exception is throw:
```
[ERROR   ] SRVE0777E: Exception thrown by application class 'org.jboss.cdi.tck.tests.context.conversation.determination.FailingServlet.service:35'
jakarta.servlet.ServletException
	at org.jboss.cdi.tck.tests.context.conversation.determination.FailingServlet.service(FailingServlet.java:35)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:587)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1258)
	at [internal classes]
```

With that servlet, I noticed that the WeldInitialListener.requestDestroyed is not called.  Based on an initial glance at WeldInitialListener. it calls HttpContextLifecycle.requestDestroyed() which seems to deactivate the conversation scope and it releases the lock. 
https://github.com/weld/core/blob/c9f77c01d295a85f786ca402e52b51ca1981e40d/modules/web/src/main/java/org/jboss/weld/module/web/servlet/HttpContextLifecycle.java#L309

```
[8/20/21, 18:20:45:558 EDT] 00000041 Servlet       2 org.jboss.weld.module.web.servlet.HttpContextLifecycle requestDestroyed WELD-000709: Destroying request com.ibm.ws.webcontainer40.srt.SRTServletRequest40@89477001
[8/20/21, 18:20:45:559 EDT] 00000041 Conversation  2 org.jboss.weld.module.web.servlet.ConversationContextActivator deactivateConversationContext WELD-000320: Cleaning up transient conversation
[8/20/21, 18:20:45:559 EDT] 00000041 Context       2 org.jboss.weld.contexts.AbstractContext destroy WELD-000201: Context org.jboss.weld.module.web.context.http.LazyHttpConversationContextImpl@4750f81d cleared
[8/20/21, 18:20:45:559 EDT] 00000041 Context       2 org.jboss.weld.contexts.beanstore.AttributeBeanStore clear WELD-000201: Context org.jboss.weld.module.web.context.beanstore.http.LazySessionBeanStore@2efa9881 cleared
[8/20/21, 18:20:45:559 EDT] 00000041 Conversation  2 org.jboss.weld.contexts.conversation.ConversationImpl unlock WELD-000314: Lock released on conversation Transient conversation
```

However,  this appears to be a general problem as the  requestDestroyed method is not called after the error in the async servlet? This problem only occurs when deferServletRequestListenerDestroyOnError is true. Therefore, it appears there may be a bug with webcontainer here. I've created  issue  #18281. 

Changing deferServletRequestListenerDestroyOnError to true for servlet 5.0 should wait until #18281 is looked into further.

**NOTE:  I haven't tested whether fixing the WeldInitialListener.requestDestroyed  call addresses the TCK test. I'm not sure where to call the listeners within webcontainer for the async dispatch call. Will need to speak to Phu about it.**  

Follow Ups:
- Test server with this app
  [ x ] Tested on Tomcat 10. Everything checks out.  

- I'll see what the servlet spec has any specifics regarding when ServletRequestListener is called. 
  [ x ]  Not much information here. I'll check again later with fresh eyes. 
  
- Verify my analysis with the CDI team. Would be good to have them double check. 
   [  ] todo

**Comment by volosied** [comment]
Updated text in previous comment to keep information together.

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer.servlet.3.1_fat

**Comment by pnicolucci** [comment]
#run-libby-bot

**Comment by volosied** [comment]
#build -- Abandoned

File: dev/com.ibm.ws.webcontainer.servlet.4.0_fat/fat/src/com/ibm/ws/fat/wc/tests/WCServerMiscTest.java [hunk 2]
```diff
@@ -42,7 +42,7 @@
 
     @BeforeClass
     public static void setUp() throws Exception {
-        server.startServer(WCServerTest.class.getSimpleName() + ".log");
+        server.startServer(WCServerMiscTest.class.getSimpleName() + ".log");
```
Comment (volosied) [inline dev/com.ibm.ws.webcontainer.servlet.4.0_fat/fat/src/com/ibm/ws/fat/wc/tests/WCServerMiscTest.java:45]: Found this error in another test.

**Comment by volosied** [comment]
#build

File: dev/com.ibm.ws.webcontainer/resources/OSGI-INF/l10n/metatype.properties [hunk 3]
```diff
@@ -146,7 +146,7 @@ emptyServletMappings.name=Return empty servlet mappings
 emptyServletMappings.desc=Toggle if you want to return an empty collection, instead of null, when no servlet mappings are added. The default value is false. The equivalent custom property in the full application server profile is com.ibm.ws.webcontainer.emptyservletmappings.
 
 deferServletRequestListenerDestroyOnError.name=Defer servlet request listener destroy
-deferServletRequestListenerDestroyOnError.desc=Toggle if you want to defer ServletRequestListener destroy when there is an error serving the request. The default value is false. The equivalent custom property in the full application server profile is com.ibm.ws.webcontainer.deferServletRequestListenerDestroyOnError.
+deferServletRequestListenerDestroyOnError.desc=Toggle if you want to defer ServletRequestListener destroy when there is an error serving the request. The default value is false, but is true for servlet 5.0 and higher. The equivalent custom property in the full application server profile is com.ibm.ws.webcontainer.deferServletRequestListenerDestroyOnError.
```
Comment (helyarp) [inline dev/com.ibm.ws.webcontainer/resources/OSGI-INF/l10n/metatype.properties]: Please replace "higher" with "later"

**Review by helyarp — COMMENTED** [review]
Suggested change

File: dev/com.ibm.ws.webcontainer/resources/OSGI-INF/l10n/metatype.properties (hunk 3 above)
Comment (volosied) [reply dev/com.ibm.ws.webcontainer/resources/OSGI-INF/l10n/metatype.properties]: Resolved.

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer.servlet.3.1_fat

**Review by helyarp — COMMENTED** [review]
Looks good!

File: dev/com.ibm.ws.webcontainer.servlet.4.0_fat/fat/src/com/ibm/ws/fat/wc/tests/WCServerPropertyTest.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                    assertTrue("The response did not contain the following String: " + expectedResponse, responseText.contains(expectedResponse));
+                } else {
+                    assertTrue("The response should have been empty! " + expectedResponse, responseText.length() == 0);
+                }
+
+            }
+        }
+    }
+
+}
```
Comment (pnicolucci) [inline dev/com.ibm.ws.webcontainer.servlet.4.0_fat/fat/src/com/ibm/ws/fat/wc/tests/WCServerPropertyTest.java]: new line please

File: dev/com.ibm.ws.webcontainer.servlet.4.0_fat/publish/servers/servlet40_jspCdi/server.xml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    
+    <!-- FAT framework expects to start with the usual port  -->
+    <httpEndpoint id="defaultHttpEndpoint"
+                  host="*"
+                  httpPort="${bvt.prop.HTTP_default}"
+                  httpsPort="${bvt.prop.HTTP_default.secure}">
+        <samesite lax="*"/>
+    </httpEndpoint>
+
+</server>
```
Comment (pnicolucci) [inline dev/com.ibm.ws.webcontainer.servlet.4.0_fat/publish/servers/servlet40_jspCdi/server.xml]: new line please

File: dev/com.ibm.ws.webcontainer.servlet.4.0_fat/test-applications/DeferServletRequestListenerDestroyOnErrorTest.war/resources/WEB-INF/web.xml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-app_4_0.xsd">
+
+  <display-name> DeferServletRequestListenerDestroyOnError Default Property Test </display-name>
+
+	<error-page>
+		<error-code>500</error-code>
+		<location>/errorPage.jsp</location>
+	</error-page>
+
+</web-app>
```
Comment (pnicolucci) [inline dev/com.ibm.ws.webcontainer.servlet.4.0_fat/test-applications/DeferServletRequestListenerDestroyOnErrorTest.war/resources/WEB-INF/web.xml]: new line please

File: dev/com.ibm.ws.webcontainer.servlet.4.0_fat/bnd.bnd [hunk 7]
```diff
@@ -52,4 +53,6 @@ tested.features:\
     httpunit:httpunit;version=1.5.4,\
     net.sf.jtidy:jtidy;version=9.3.8,\
     org.apache.httpcomponents.client5:httpclient5;version=5.0.2,\
-    org.apache.httpcomponents.core5:httpcore5;version=5.0.2
+    org.apache.httpcomponents.core5:httpcore5;version=5.0.2,\
+    com.ibm.websphere.javaee.cdi.2.0;version=latest,\
+    com.ibm.ws.logging.core;version=latest
```
Comment (pnicolucci) [inline dev/com.ibm.ws.webcontainer.servlet.4.0_fat/bnd.bnd]: Is this logging entry necessary?

Comment (volosied) [reply dev/com.ibm.ws.webcontainer.servlet.4.0_fat/bnd.bnd]: Turns out it's not. I've removed it.

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer.servlet.3.1_fat

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer.servlet.3.1_fat

**Review by pnicolucci — APPROVED** [review]

**Comment by skasund** [comment]
L2 message review not required

**Comment by volosied** [comment]
#run-libby-bot

**Pull request merged**